### PR TITLE
Alerts System and UI

### PR DESCRIPTION
**Pull request by chairbender**
Adds the necessary stuff to achieve the alerts UI as shown in the injazz mockups.

- Alerts are now defined as prototypes and can be shown / hidden by id. They accept formatted title/description. Category can be used to have alerts which automatically replace each other, such as hot / cold or low/high pressure.
- added alert prototypes for all of our current alerts. I don't intend for the title/descriptions to be final, some could do with some more color highlighting or revising, they are mostly based on tg.
- Alerts have a defined order which is also done via a prototype
- Alert toolbar now has a background and can support arbitrary many tooltips. It expands left when it gets too tall. It should properly size itself even when you change the window size or UIScale.
- Alerts have a nicer tooltip popup based on their name and description. If the alert has a cooldown, such as stun, it will also show the total duration.
- Alert OnClick behavior is defined in code when showing the alert on the server side.
- Alert data is sent to client in a compact, encoded format (no strings!)
- adds showalert /  clearalert command
- I intentionally left in some debug alerts (debug1 - debug6). A lot of our alerts will disappear automatically, so debug alerts can be used to easily see what happens when you have a lot of alerts going on. For example to create a lot of alerts to see how the UI handles it, `showalert Debug1 -1` and repeat that for Debug2-6 to fill up the bar.
- The alerts toolbar and tooltips have been styled to try to match the injazz mockups as closely as possible, down to exact colors, fonts, etc...this included adding some new panel textures.
- Adds test coverage for this stuff, including multiple unit tests and an integration test

Part of #272 

Requires space-wizards/RobustToolbox#1395

### CODE REVIEW

**Comment by metalgearsloth** [comment]
Got a screenshot?

**Comment by chairbender** [comment]
image

File: Content.Server/GameObjects/Components/Mobs/ServerAlertsComponent.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            {
+                weightlessSystem.RemoveAlert(this);
+            }
+            else
+            {
+                Logger.WarningS("alert", "weightlesssystem not found");
+            }
+
+            base.OnRemove();
+        }
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Mobs/ServerAlertsComponent.cs:48]: Maybe use eventbus instead?

File: Content.Shared/Alert/AlertPrototype.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    public class AlertPrototype : IPrototype
+    {
+        private string _id;
+        private string _icon;
+        private FormattedMessage _name;
+        private FormattedMessage _description;
+        private string _category;
+        private AlertKey _alertKey;
+        private short _maxSeverity;
+        private short _minSeverity;
```
Comment (metalgearsloth) [inline Content.Shared/Alert/AlertPrototype.cs:24]: Could just make these properties with public getters and private setters.

File: Content.Shared/Alert/AlertPrototype.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+        /// <summary>
+        /// Unique identifier used to reference this state in code
+        /// </summary>
+        [ViewVariables]
+        public string ID => _id;
+
+        /// <summary>
+        /// Path to the icon (png) or  to show in alert bar. If severity levels are supported,
```
Comment (metalgearsloth) [inline Content.Shared/Alert/AlertPrototype.cs:33]: or...

File: Content.Shared/GameObjects/Components/Mobs/SharedAlertsComponent.cs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+        /// <summary>
+        /// Creates a new array containing all of the current alert states.
+        /// </summary>
+        /// <returns></returns>
+        protected AlertState[] CreateAlertStatesArray()
+        {
+            if (_alerts.Count == 0) return NO_ALERTS;
+            var states = new AlertState[_alerts.Count];
+            // because I don't trust LINQ
```
Comment (metalgearsloth) [inline Content.Shared/GameObjects/Components/Mobs/SharedAlertsComponent.cs:87]: me too buddy, me too.

File: Content.Shared/GameObjects/Components/Mobs/SharedAlertsComponent.cs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        {
+            Directed = true;
+            EncodedAlert = encodedAlert;
+        }
+    }
+
+    [Serializable, NetSerializable]
+    public struct AlertState
+    {
+        public int AlertEncoded;
```
Comment (metalgearsloth) [inline Content.Shared/GameObjects/Components/Mobs/SharedAlertsComponent.cs:249]: Could you use byte or ushort? pedantic but minor savings.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
IMO the effects background should be opaque like the mockup and the text red as well. Also the alerts should be an enum instead just to avoid typo issues only being detected at runtime. This also means you can remove the string checks for alert names as well.

Otherwise looks really good.

File: Content.Server/GameObjects/Components/Mobs/ServerAlertsComponent.cs (hunk 1 above)
Comment (chairbender) [reply Content.Server/GameObjects/Components/Mobs/ServerAlertsComponent.cs:48]: this is not something new by me, this was how it was already implemented in the existing ServerStatusEffectsComponent which I have renamed to ServerAlertsComponent

File: Resources/Prototypes/Alerts/alerts.yml [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+- type: alert
+  id: humancrit
+  category: health
+  icon: /Textures/Interface/StatusEffects/Human/humancrit-0.png
+  name: "[color=red]Critical Condition[/color]"
+  description: "You're severely injured and unconscious."
+
+- type: alert
+  id: humandead
```
Comment (exp111) [inline Resources/Prototypes/Alerts/alerts.yml:95]: Shouldn't those be camelCase?

Comment (Zumorica) [reply Resources/Prototypes/Alerts/alerts.yml:95]: Let me quote the official naming guidelines for YAML:
>`PascalCase` is used for IDs and component names. Everything else uses `camelCase`.

**Comment by Zumorica** [comment]
That UI looks really, really good! Nice job.

**Comment by chairbender** [comment]
thanks for the feedback, it's waaay better now using enums

@metalgearsloth ready for another look

File: Content.Shared/Alert/AlertPrototype.cs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            return IconPath.Substring(0, ext) + severity + IconPath.Substring(ext, IconPath.Length - ext);
+        }
+    }
+
+    /// <summary>
+    /// Key for an alert which is unique (for equality and hashcode purposes) w.r.t category semantics.
+    /// I.e., entirely defined by the category, if a category was specified, otherwise
+    /// falls back to the id.
+    /// </summary>
+    [Serializable]
```
Comment (metalgearsloth) [inline Content.Shared/Alert/AlertPrototype.cs:143]:
```suggestion
    [Serializable, NetSerializable]
```

**Review by metalgearsloth — COMMENTED** [review]
Just 2 minor things:

1. Could rename StatusEffects folder to Alerts
2. Just that comment about NetSerializable.

**Comment by chairbender** [comment]
fixed and renamed

**Pull request merged**